core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,197 @@
+    */
+
+    var _onFinishedLoadingImage = function(imageFilename, onLoadImageResolve) {
+      if (_recentlyRequestedImageFilenames.indexOf(imageFilename) !== -1) {
+        _getUrlUsingFileInCache(imageFilename, onLoadImageResolve);
+      }
+    };
+
+    var _getImageFilenamesInBfsOrder = function(sourceStateName) {
+      var stateNamesInBfsOrder = (
+        ComputeGraphService.computeBfsTraversalOfStates(
+          _exploration.getInitialState().name,
+          _exploration.getStates(),
+          sourceStateName));
+      var imageFilenames = [];
+
+      stateNamesInBfsOrder.forEach(function(stateName) {
+        var state = _states.getState(stateName);
+        ExtractImageFilenamesFromStateService.getImageFilenamesInState(state).
+          forEach(function(filename) {
+            imageFilenames.push(filename);
+          });
+      });
+      return imageFilenames;
+    };
+
+    var _loadImage = function(imageFilename) {
+      AssetsBackendApiService.loadImage(
+        ExplorationContextService.getExplorationId(), imageFilename
+      ).then(function(loadedImage) {
+        for (var i = 0;
+          i < _filenamesOfImageCurrentlyDownloading.length; i++) {
+          if (_filenamesOfImageCurrentlyDownloading[i] ===
+              loadedImage.filename) {
+            _filenamesOfImageCurrentlyDownloading.splice(i, 1);
+            break;
+          }
+        }
+        if (_filenamesOfImageToBeDownloaded.length > 0) {
+          var nextImageFilename = _filenamesOfImageToBeDownloaded.shift();
+          _filenamesOfImageCurrentlyDownloading.push(nextImageFilename);
+          _loadImage(nextImageFilename);
+        }
+        if (_imageLoadedCallback[loadedImage.filename]) {
+          var onLoadImageResolve = (
+            (_imageLoadedCallback[loadedImage.filename]).resolveMethod);
+          _onFinishedLoadingImage(loadedImage.filename, onLoadImageResolve);
+          _imageLoadedCallback[loadedImage.filename] = null;
+        }
+      });
+    };
+
+    var _kickOffImagePreloader = function(sourceStateName) {
+      _filenamesOfImageToBeDownloaded = (
+        _getImageFilenamesInBfsOrder(sourceStateName));
+      while (_filenamesOfImageCurrentlyDownloading.length <
+        MAX_NUM_IMAGE_FILES_TO_DOWNLOAD_SIMULTANEOUSLY &&
+          _filenamesOfImageToBeDownloaded.length > 0) {
+        var imageFilename = _filenamesOfImageToBeDownloaded.shift();
+        _filenamesOfImageCurrentlyDownloading.push(imageFilename);
+        _loadImage(imageFilename);
+      }
+    };
+
+    var _cancelPreloading = function() {
+      AssetsBackendApiService.abortAllCurrentImageDownloads();
+      _filenamesOfImageCurrentlyDownloading = [];
+    };
+
+    var _onStateChange = function(stateName) {
+      if (stateName !== _exploration.getInitialState().name) {
+        _imageLoadedCallback = {};
+        var imageFilenamesInState = [];
+        var imageFilenamesInStateCurrentlyBeingRequested = [];
```

Comment:
Do you need arrays? Just number of files might be enough in this case.

Reply:
Done!